tests/components/sma/test_config_flow.py

```diff
@@ -59,90 +84,120 @@ async def test_form_cannot_connect(hass: HomeAssistant) -> None:
-    assert result["errors"] == {"base": "cannot_connect"}
+    assert result["errors"] == {"base": error}
     assert len(mock_setup_entry.mock_calls) == 0
 
 
-async def test_form_invalid_auth(hass: HomeAssistant) -> None:
-    """Test we handle invalid auth error."""
+async def test_form_already_configured(hass: HomeAssistant, mock_config_entry) -> None:
+    """Test starting a flow by user when already configured."""
+    mock_config_entry.add_to_hass(hass)
+
     result = await hass.config_entries.flow.async_init(
         DOMAIN, context={"source": SOURCE_USER}
     )
 
     with (
-        patch("pysma.SMA.new_session", side_effect=SmaAuthenticationException),
+        patch("pysma.SMA.new_session", return_value=True),
+        patch("pysma.SMA.device_info", return_value=MOCK_DEVICE),
+        patch("pysma.SMA.close_session", return_value=True),
         _patch_async_setup_entry() as mock_setup_entry,
     ):
         result = await hass.config_entries.flow.async_configure(
             result["flow_id"],
             MOCK_USER_INPUT,
         )
 
-    assert result["type"] is FlowResultType.FORM
-    assert result["errors"] == {"base": "invalid_auth"}
+    assert result["type"] is FlowResultType.ABORT
+    assert result["reason"] == "already_configured"
     assert len(mock_setup_entry.mock_calls) == 0
 
 
-async def test_form_cannot_retrieve_device_info(hass: HomeAssistant) -> None:
-    """Test we handle cannot retrieve device info error."""
+async def test_dhcp_discovery(hass: HomeAssistant) -> None:
+    """Test we can setup from dhcp discovery."""
     result = await hass.config_entries.flow.async_init(
-        DOMAIN, context={"source": SOURCE_USER}
+        DOMAIN,
+        context={"source": SOURCE_DHCP},
+        data=DHCP_DISCOVERY,
     )
 
+    assert result["type"] is FlowResultType.FORM
+    assert result["step_id"] == "discovery_confirm"
+
     with (
         patch("pysma.SMA.new_session", return_value=True),
-        patch("pysma.SMA.read", side_effect=SmaReadException),
+        patch("pysma.SMA.device_info", return_value=MOCK_DEVICE),
         _patch_async_setup_entry() as mock_setup_entry,
     ):
         result = await hass.config_entries.flow.async_configure(
             result["flow_id"],
-            MOCK_USER_INPUT,
+            MOCK_DHCP_DISCOVERY_INPUT,
         )
+        await hass.async_block_till_done()
+
+    assert result["type"] is FlowResultType.CREATE_ENTRY
+    assert result["title"] == MOCK_DHCP_DISCOVERY["host"]
+    assert result["data"] == MOCK_DHCP_DISCOVERY
+
+    assert len(mock_setup_entry.mock_calls) == 1
 
-    assert result["type"] is FlowResultType.FORM
-    assert result["errors"] == {"base": "cannot_retrieve_device_info"}
-    assert len(mock_setup_entry.mock_calls) == 0
 
+async def test_dhcp_already_configured(hass: HomeAssistant, mock_config_entry) -> None:
+    """Test starting a flow by dhcp when already configured."""
+    mock_config_entry.add_to_hass(hass)
```

Comment:
We should already be able to abort when the DHCP flow starts, since we know the serial number and thus already can abort